src/Illuminate/Database/Eloquent/FactoryBuilder.php

```diff
@@ -131,7 +132,22 @@ protected function makeInstance(array $attributes = [])
 
             $definition = call_user_func($this->definitions[$this->class][$this->name], $this->faker, $attributes);
 
-            return new $this->class(array_merge($definition, $attributes));
+            $evaluated = $this->evaluateClosures(array_merge($definition, $attributes));
+
+            return new $this->class($evaluated);
         });
     }
+
+    /**
+     * Delay closures evaluation after merging to avoid duplication.
+     *
+     * @param  array  $attributes
+     * @return array
+     */
+    protected function evaluateClosures(array $attributes)
+    {
+        return array_map(function ($attribute) use ($attributes) {
+            return $attribute instanceof Closure ? $attribute($attributes) : $attribute;
```

Comment:
Graham I think this is a real argument for for not using value. I have a few use cases that depend on knowing what exactly is being created in order to give each attribute a meaningful value.

Devin Dombrowski
dddesigns3@gmail.com




> On Feb 10, 2016, at 7:40 AM, Alessandro Benoit <notifications@github.com> wrote:
> 
> In src/Illuminate/Database/Eloquent/FactoryBuilder.php <https://github.com/laravel/framework/pull/12231#discussion_r52473981>:
> 
> >          });
> >      }
> > +
> > +    /**
> > +     * Delay closures evaluation after merging to avoid duplication.
> > +     *
> > +     * @param  array  $attributes
> > +     * @return array
> > +     */
> > +    protected function evaluateClosures(array $attributes)
> > +    {
> > +        return array_map(function ($attribute) use ($attributes) {
> > +            return $attribute instanceof Closure ? $attribute($attributes) : $attribute;
> @GrahamCampbell <https://github.com/GrahamCampbell> Using value won't allow to pass the evaluated definitions as the closure attribute, which is essential to work with complex relationships:
> 
> $factory->define(App\Post::class, function () {
>     return [
>         'user_id' => function () {
>             return factory(App\User::class)->create()->id;
>         },
>         'user_type' => function (array $evaluated) {
>             return App\User::find($evaluated['user_id'])->type;
>         }
>     ];
> });
> —
> Reply to this email directly or view it on GitHub <https://github.com/laravel/framework/pull/12231/files#r52473981>.
>